### ISSUE TITLE
Adding triadic analysis from sociology

### ISSUE THREAD

**Issue by bakerwho** [networkx/networkx#3727]
I'm proposing to augment NetworkX with functionality for triadic analysis from sociology. I welcome comments and suggestions about how to organize this as a potential PR. This is for a course project, but I would love to actually contribute my code to NetworkX.

##### Introduction to sociological triads
Sociology has long dealt with the concept of triadic interactions. As conceptualized by late 20th century sociologist Georg Simmel, a dyad (or interaction between two parties) is simple - it can be positive or negative, strong or weak. However, studying triadic interactions unearths a significant degree of complexity. A simple example is Simmel's idea of triadic closure - that a strong positive relation between A and B, and a strong positive relation between B and C, imply that there should be a strong positive relation between A and C. Consider a set of three friends, or three members of a family, or three allies in global politics - the rule is strangely pervasive.

This has led to several notions and theories of behavior based of these triads, constituting a type of analysis that is now known as structural interactionism. A huge body of work analyses various mechanisms and patterns that can be identified from simple triadic analysis.

##### Python for triadic analysis
NetworkX has limited provisions for triadic analysis. The only functionality I could find was `triadic_census()`, which simply counts the number of triads in a given directed graph (see [here](https://networkx.github.io/documentation/networkx-2.2/reference/algorithms/triads.html)).

Triadic analysis will likely get very expensive for large networks. However, the goal is first to have a proof of concept that it can be done in conjunction with NetworkX.

##### Proposal
I propose to extend and adapt NetworkX code to simple sociological triadic analysis. I can begin by adding the following functions on a graph `G` (to be expanded):
 - `G.get_triadic_closures()`  - list of closed triads within the graph
 - `G.get_focal_closures(attr_name)` - list the triads which are closed around a node attribute `attr_name`
 - `G.get_stable_triads(stability_func)` - count triads that fulfill a given criterion of stability (a function that takes the graph, 3 nodes, and a criterion on them, eg: focal closure). The idea is to count unstable triads. These could be expected to stabilize either by formation or deletion of existing edges

The following are more crudely thought-out functions for sociological analysis that can be implemented using NetworkX:
 - `G.predict_edge_deletion()`
 - `G.predict_edge_formation()`
 - `G.simulate_triadic_mechanism()`

**Comment by ericmjl** [comment]
@bakerwho I'm super in favour of this! Though I do also want to hear your thoughts: what about making it part of the `nx.___` API as a function, rather than as a class method? 

As far as I have seen, there's an (admittedly fuzzy) distinction between what goes on as a class method and what becomes a networkx function, but I sort of see it as "elementary graph properties" become class methods, while "less elementary graph properties" (e.g. triadic closures) would be functions.

It would look like:

- `nx.triadic_closures(G)`
- `nx.focal_closures(G)`
- `nx.stable_triads(G)`

We could then dispense away with the `get`ters, which is less Pythonic than simply naming a function by the logical thing it's supposed to return.

What do you think?

**Comment by bakerwho** [comment]
@ericmjl sounds good! I appreciate the design discussion to help guide me on this. 

Would it then be ideal for me to add these functions to `networkx/algorithms/triads.py`? This is the where `triadic_census()` is currently defined.

**Comment by ericmjl** [comment]
I think we should pull in @dschult into the discussion - he would have a much better idea than I would! :smile:

**Comment by dschult** [comment]
Yes, please add them to ```triads.py``` along with tests in the ```tests/test_triads.py``` module. 

The PEP8 style guidelines can be tested locally using pycodestyle (pip install pycodestyle). 

In addition the doc_strings for each function should start with a single line description with <80 chars (sphinx uses this for short descriptions for links to a function in the docs), then a blank line and then a paragraph description with more details.

Thanks!!

**Comment by bakerwho** [comment]
@dschult and @ericmjl, thanks for the response and interest!

For graphs often studied in sociology, it might be nice to have a dict of **lists of actual triads by type** `{'003':(subgraph1, subgraph2), '012':...}`. The added benefit is that these can be used to simulate dynamical processes on these triads. As a simple example, triadic closure would dictate that if A<->B and B<->C and A<-x->C,  then the edge A<->B should be generated. Having lists of triads by type would simplify the process of implementing these rules based on triad type.

I was thinking of writing `nx.triads_by_type()` to do this. This would also be helpful to functions like `nx.triadic_closures(G)`, `nx.focal_closures(G)`, `nx.stable_triads(G)` that we discussed above.

`nx.triangles()` could potentially be useful for this functionality, but is implemented only for undirected graphs. In any case, bruteforcing would be O(n^3) complex (I think). If bruteforce is the only way, we could perhaps have cap on the order of input graphs, or a warning.

Do you think it might be worthwhile to have `nx.triads_by_type()`?

**Comment by bakerwho** [comment]
@dschult and @ericmjl, bump on this discussion after Thanksgiving! Happy holidays and hoping to get your thoughts on this!

**Comment by ericmjl** [comment]
Ahhh the turkey! 

@bakerwho those sound like good ideas! I'm going to defer to @dschult, and maybe also pull in @MridulS and @jarrodmillman on the thread to see if they have other thoughts.

In terms of "actual work", I would recommend submitting a PR that isn't too large so that it's easier to review. A suggestion you I can offer (that you're free to reject) is to first submit `nx.triadic_closures(G)` and its implementation, docs, and tests. Happy to help with the review, as long as you tag me in there :smile:.

**Comment by dschult** [comment]
This sounds good -- I think the first step might be a ```nx.triadic_closures``` function PR (that ignores triad type? -- not sure how the triad type affects things).
:)

**Comment by bakerwho** [comment]
@dschult @ericmjl I've made some progress on this. For one, `nx.triadic_closures` would need a list/generator of all triads.

Triadic closure is simply transitivity of edges (if A<->B and A<->C, then B<->C). Each of the 16 triad types can be identified as respecting triadic closure or not.

Triadic closure is also a fairly simple and rudimentary sociological concept that is treated as a 'useful simplification'. So I figured it would make the most sense to build this function on top of `nx.triads_by_type()` which returns a dict of triad types as keys and lists of triads as values.
